share/man/man4/man4.sun3x/kbd.4
.\"	$NetBSD: kbd.4,v 1.1 1997/12/24 08:42:14 jeremy Exp $
.\"
.\" Copyright (c) 1996 The NetBSD Foundation, Inc.
.\" All rights reserved.
.\"
.\" This code is derived from software contributed to The NetBSD Foundation
.\" by Paul Kranenburg.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"        This product includes software developed by the NetBSD
.\"        Foundation, Inc. and its contributors.
.\" 4. Neither the name of The NetBSD Foundation nor the names of its
.\"    contributors may be used to endorse or promote products derived
.\"    from this software without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
.\" ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
.\" TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS 
.\" BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd June 22, 1996
.Dt KBD 4 sun3
.Os
.Sh NAME
.Nm kbd
.Nd workstation keyboard
.Sh SYNOPSIS
.Cd "pseudo-device kbd"
.Sh DESCRIPTION
The
.Nm
driver provides an interface to the workstation console keyboard.
.Dq type 2 ,
.Dq type 3 ,
and
.Dq type 4
keyboards are supported. All types generate keycodes encoding the
key identity and motion (up or down) as the keys are pressed and released.
The
.Nm
driver either passes the keycodes to an application as they come in, or
translates them into ASCII characters first according to a set of built-in
tables.
If the keyboard is configured as the device
to be used for system console input
.Pq see Xr openprom 4 ,
it will be internally connected to the
.Pa /dev/console
device special file, which can be used as a
.Xr tty 4
device.
The device special file
.Pa /dev/kbd
is used to get direct access to the keyboard input stream. The following
ioctl's are supported (mostly just enough to keep the X11 server going):
.Bl -tag -width KIOCSDIRECT
.It KIOCTRANS
Set translation mode. The argument is of type
.Fa "int *" ,
the only value supported is
.Dv TR_UNTRANS_EVENT .
.It KIOCGTRANS
Get translation mode. The argument is of type
.Fa "int *" .
.Dv TR_UNTRANS_EVENT
is always returned.
.It KIOCGETKEY
Fill in old-style key station translation. The argument is of type
.Fa "struct okiockey *" .
.It KIOCCMD
Send a command to the keyboard. The argument is of type
.Fa "int *" ,
and can have one of the following values:
.Bl -tag -width KBD_CMD_NOCLICK
.It KBD_CMD_BELL
Start the keyboard beeper.
.It KBD_CMD_NOBELL
Stop the keyboard beeper.
.It KBD_CMD_CLICK
Instruct the keyboard to make extra noise when touching keys.
.It KBD_CMD_NOCLICK
Instruct the keyboard to.. oh well..
.El
.It KIOCTYPE
Get keyboard type. The argument is of type
.Fa "int *" ,
in which one of the values
.Dv KB_SUN2 ,
.Dv KB_SUN3
or
.Dv KB_SUN4
will be returned.
.It KIOCSDIRECT
Route the keyboard input stream through the (SunOS) compatible event module.
The argument is of type
.Fa "int *" ,
a non-zero value will put the driver into
.Dq event
mode, while a value of zero will make it return to
.Dq ASCII translation
mode.
.It KIOCSKEY
Set key station translation. The argument is of type
.Fa "struct kiockey *" 
.Pq see Pa /usr/include/machine/kbio.h No for more details .
.It KIOCGKEY
Get key station translation. The argument is of type
.Fa "struct kiockey *" .
.It KIOCLAYOUT
Get keyboard layout
.Pq Do type 4 Dc only .
The argument is of type
.Fa "int *" ,
in which the uninterpreted result of the
.Dv KBD_CMD_GLAYOUT
keyboard command is returned
.Po on Dv KBDUN4
type keyboards this will be the setting of a DIP switch bank
.Pc .
.It KIOCSLED
Set LED state
.Pq Do type 4 Dc only .
The argument is of type
.Fa "char *" ,
and is the inclusive OR of the following flags:
.Pp
.Bl -tag -width LED_SCROLL_LOCK -compact
.It LED_NUM_LOCK
.It LED_COMPOSE
.It LED_SCROLL_LOCK
.It LED_CAPS_LOCK
.El
.Pp
Each of these flags turn on the LED in the obvious key.
.It KIOCGLED
Get LED state
.Pq Do type 4 Dc only .
The argument is of type
.Fa "char *" ,
in which the current LED state is returned.
.El
.Sh SEE ALSO
.Xr ms 4
.Sh BUGS
.Nm
is hardwired to the built-in
.Em zs1
serial port.


sys/arch/x68k/dev/kbd.c
/*	$NetBSD: kbd.c,v 1.1 1996/05/05 12:17:06 oki Exp $	*/

/*
 * Copyright (c) 1982, 1986, 1990 The Regents of the University of California.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include "ite.h"
#include "bell.h"

#if NITE > 0
#include <sys/param.h>
#include <sys/systm.h>
#include <sys/device.h>
#include <sys/ioctl.h>
#include <sys/tty.h>
#include <sys/proc.h>
#include <sys/conf.h>
#include <sys/file.h>
#include <sys/uio.h>
#include <sys/kernel.h>
#include <sys/syslog.h>

#include <machine/cpu.h>

#include <x68k/dev/itevar.h>
#include <x68k/x68k/iodevice.h>

/* for sun-like event mode, if you go thru /dev/kbd. */
#include <x68k/dev/event_var.h>
#include <machine/kbio.h>
#include <machine/kbd.h>
#include <machine/vuid_event.h>

struct kbd_softc {
	int k_event_mode;  	 /* if true, collect events, else pass to ite */
	struct evvar k_events; /* event queue state */
} kbd_softc;

void	kbdattach __P((int));

/*
 * Called from main() during pseudo-device setup.  If this keyboard is
 * the console, this is our chance to open the underlying serial port and
 * send a RESET, so that we can find out what kind of keyboard it is.
 */
void
kbdattach(kbd)
	int kbd;
{
}

/* definitions for x68k keyboard encoding. */
#define KEY_CODE(c)  ((c) & 0x7f)
#define KEY_UP(c)    ((c) & 0x80)

void
kbdenable ()
{
	int s = spltty();

	sysport.keyctrl = 8;	/* send character from keyboard enable */
	mfp.iera &= (~0x11);	/* MPSC RBF, Timer-B interrupt disable */
	mfp.tbcr = MFP_TIMERB_RESET | MFP_TIMERB_STOP;    /* Timer-B stop */
	mfp.tbdr = 13;		/* Timer-B 38400 Hz clock (interrupt 76800Hz) */
	mfp.tbcr = 1;		/* Timer-B deray mode, prescale 1/4 */
	mfp.ucr = MFP_UCR_CLKX16 | MFP_UCR_RW_8 | MFP_UCR_ONESB;
	mfp.rsr = MFP_RSR_RE;	/* USART receive enable */
	mfp.iera |= 0x11;	/* MPSC RBF, Timer-B interrupt enable */

	while(!(mfp.tsr & MFP_TSR_BE)) ;
	mfp.udr = 0x49;     /* send character from keyboard enable */
	kbdled = 0;	    /* all keyboard LED turn off. */
	kbd_setLED();

	if (!(sysport.keyctrl & 8))
		printf("WARNING: no connected keyboard\n");
	kbd_softc.k_event_mode = 0;
	kbd_softc.k_events.ev_io = 0;
	splx(s);
}

int
kbdopen(dev, flags, mode, p)
	dev_t dev;
	int flags, mode;
	struct proc *p;
{
	int s, error;

	if (kbd_softc.k_events.ev_io)
		return (EBUSY);
	kbd_softc.k_events.ev_io = p;
	ev_init(&kbd_softc.k_events);
	return (0);
}

int
kbdclose(dev, flags, mode, p)
	dev_t dev;
	int flags, mode;
	struct proc *p;
{
	/* Turn off event mode, dump the queue */
	kbd_softc.k_event_mode = 0;
	ev_fini(&kbd_softc.k_events);
	kbd_softc.k_events.ev_io = NULL;
	return (0);
}

int
kbdread(dev, uio, flags)
	dev_t dev;
	struct uio *uio;
	int flags;
{
	return ev_read(&kbd_softc.k_events, uio, flags);
}

/* this routine should not exist, but is convenient to write here for now */
int
kbdwrite(dev, uio, flags)
	dev_t dev;
	struct uio *uio;
	int flags;
{
	return EOPNOTSUPP;
}

int
kbdioctl(dev, cmd, data, flag, p)
	dev_t dev;
	u_long cmd;
	caddr_t data;
	int flag;
	struct proc *p;
{
	register struct kbd_softc *k = &kbd_softc;
	int cmd_data;

	switch (cmd) {
	case KIOCTRANS:
		if (*(int *)data == TR_UNTRANS_EVENT)
			return (0);
		break;

	case KIOCGTRANS:
		/*
		 * Get translation mode
		 */
		*(int *)data = TR_UNTRANS_EVENT;
		return (0);

	case KIOCSDIRECT:
		k->k_event_mode = *(int *)data;
		return (0);

	case KIOCCMD:
		cmd_data = *(int *)data;
		return kbd_send_command(cmd_data);

	case KIOCSLED:
		kbdled = *(char *)data;
		kbd_setLED();
		return (0);

	case KIOCGLED:
		*(char *)data = kbdled;
		return (0);

	case KIOCSBELL:
#if NBELL > 0
		/* XXX - so tricky! */
		return opm_bell_setup((struct kbiocbell *)data);
#else
		return (0);	/* allways success */
#endif

	case FIONBIO:		/* we will remove this someday (soon???) */
		return (0);

	case FIOASYNC:
		k->k_events.ev_async = *(int *)data != 0;
		return (0);

	case TIOCSPGRP:
		if (*(int *)data != k->k_events.ev_io->p_pgid)
			return (EPERM);
		return (0);

	default:
		return (ENOTTY);
	}

	/*
	 * We identified the ioctl, but we do not handle it.
	 */
	return (EOPNOTSUPP);		/* misuse, but what the heck */
}

int
kbdselect(dev, rw, p)
	dev_t dev;
	int rw;
	struct proc *p;
{
	return ev_select (&kbd_softc.k_events, rw, p);
}

#define KBDBUFMASK 63
#define KBDBUFSIZ 64
static u_char kbdbuf[KBDBUFSIZ];
static int kbdputoff = 0;
static int kbdgetoff = 0;

int
kbdintr()
{
	u_char c, in;
	struct kbd_softc *k = &kbd_softc;
	struct firm_event *fe;
	int put;

	c = in = mfp.udr;

	/* if not in event mode, deliver straight to ite to process key stroke */
	if (! k->k_event_mode) {
		kbdbuf[kbdputoff++ & KBDBUFMASK] = c;
		setsoftkbd();
		return;
	}

	/* Keyboard is generating events.  Turn this keystroke into an
	   event and put it in the queue.  If the queue is full, the
	   keystroke is lost (sorry!). */
  
	put = k->k_events.ev_put;
	fe = &k->k_events.ev_q[put];
	put = (put + 1) % EV_QSIZE;
	if (put == k->k_events.ev_get) {
		log(LOG_WARNING, "keyboard event queue overflow\n"); /* ??? */
		return;
	}
	fe->id = KEY_CODE(c);
	fe->value = KEY_UP(c) ? VKEY_UP : VKEY_DOWN;
	fe->time = time;
	k->k_events.ev_put = put;
	EV_WAKEUP(&k->k_events);
}

kbdsoftint()
{
	int s = splhigh();

	while(kbdgetoff < kbdputoff)
		ite_filter(kbdbuf[kbdgetoff++ & KBDBUFMASK], ITEFILT_TTY);
	kbdgetoff = kbdputoff = 0;

	splx(s);
}

int
kbd_bell(mode)
	int mode;
{
#if NBELL > 0
	if (mode)
		opm_bell_on();
	else
		opm_bell_off();
#endif
}

int
kbdgetcn()
{
	int s = spltty();
	u_char ints, c, in;

	ints = mfp.iera;

	mfp.iera &= 0xef;
	mfp.rsr |= 1;
	while (!(mfp.rsr & MFP_RSR_BF))
		asm("nop");
	in = c = mfp.udr;

	mfp.iera = ints;
	splx (s);

	return c;
}

kbd_setLED()
{
	while(!(mfp.tsr & MFP_TSR_BE)) ;
        mfp.udr = ~kbdled | 0x80;
}

int
kbd_send_command(cmd)
	int cmd;
{
	switch (cmd) {
	case KBD_CMD_RESET:
		/* XXX */
		return 0;

	case KBD_CMD_BELL:
		kbd_bell(1);
		return 0;

	case KBD_CMD_NOBELL:
		kbd_bell(0);
		return 0;

	default:
		return ENOTTY;
	}
}

#endif
